salt/pillar/nsot.py

```diff
@@ -0,0 +1,181 @@
+# -*- coding: utf-8 -*-
+'''
+A module that adds data to the Pillar structure from the NSoT API.
+
+.. versionadded:: Neon
+
+Configuring the NSoT ext_pillar
+---------------------------------
+
+The following fields are required:
+
+.. code-block:: yaml
+
+    ext_pillar:
+      - nsot:
+        api_url: http://nsot_url.com/api/
+        email: 'user@site.com'
+        secret_key: abc123
+
+.. note::
+
+    Note the trailing slash on the api_url field. Although NSoT gives the option
+    of whether or not to use the tailing slash, this module assumes that the
+    APPEND_SLASH option is set to True (the default)
+
+The following fields are optional:
+
+replace_dot: ``str``
+    This is in case a minion_id contains a dot. Because NSoT
+    doesn't allow hostnmes with dots, this option allows you to look up a device
+    with this character replacing dots. For example, with this option set to
+     '-', the minion rtr1.lax1 would be looked up in NSoT as rtr1-lax1.
+
+all_device_info: ``list``
+    This is a list of minions that will receive all device
+    information in NSoT. For example, if this list contained the entry
+    'test_min', then the minion 'test_min' would have all devices in NSoT under
+    pillar['nsot']['devices']
+
+minion_regex: ``list``
+    This is in case you only want certain minions being looked
+    up in NSoT. For example, if you have 500 minions on a single master, you may
+    not want all of them hitting the NSoT endpoint. With this list of regexes,
+    only minions matching the regex will be queried.
+
+Here's an example config with all options:
+
+.. code-block:: yaml
+
+    ext_pillar:
+      - nsot:
+        api_url: http://nsot_url.com/api/
+        email: 'user@site.com'
+        secret_key: abc234
+        replace_dot: '-'
+        all_device_info:
+          - minion_1
+          - minion_2
+        minion_regex:
+          - 'rtr*'
+          - 'sw*'
+          - '^router*'
+
+'''
+
+from __future__ import absolute_import, print_function, unicode_literals
+import logging
+import re
+
+# Import Salt libs
+import salt.utils.http
+
+log = logging.getLogger(__name__)
+
+
+def _get_token(url, email, secret_key):
+    '''
+    retrieve the auth_token from nsot
+
+    :param url: str
+    :param email: str
+    :param secret_key: str
+    :return: str
+    '''
+    url += 'authenticate/'
+    data_dict = {"email": email, "secret_key": secret_key}
+    query = salt.utils.http.query(url, data=data_dict, method='POST',
+                                  decode=True)
+    error = query.get('error')
+    if error:
+        log.error('can\'t get auth_token from nsot! reason: %s', error)
+        return False
+    else:
+        log.debug('successfully obtained token from nsot!')
+        return query['dict']['auth_token']
```

Comment:
Unless you're _really sure_ that this structure never changes and _always_ contains these keys, use `.get` instead:

```python
return query.get('dict', {}).get('auth_token', False)
```

 But then debug should also say "oops, structure is different, no token has been found"